jsonschema_extractor/attrs_extractor.py

```diff
@@ -37,33 +37,65 @@ def extract(cls, extractor, typ):
     def _extract_attribute(cls, extractor, attribute):
         is_required = attribute.default is attr.NOTHING
 
-        schema = None
         if "jsonschema" in attribute.metadata:
             schema = attribute.metadata["jsonschema"]
-        elif attribute.type is not None:
-            schema = extractor.extract(attribute.type)
         else:
+            schema = cls._extract_attribute_schema_by_type(extractor, attribute) or {}
             for validator in _iterate_validator(attribute.validator):
-                if isinstance(validator, _InstanceOfValidator):
-                    schema = extractor.extract(validator.type)
+                if isinstance(validator, _InValidator):
+                    if isinstance(validator.options, (EnumMeta, Enum)):
+                        options = _enum_to_values(validator.options)
+                    else:
+                        options = validator.options
 
-        if schema is None:
+                    # Sorting the list for constituency
+                    schema['enum'] = sorted(list(options))
+
+        if not schema:
             raise UnextractableSchema(
-                "all attributes must have an 'InstanceOfValidator'. attribute {0} does not.".format(attribute)
+                "all attributes must have an 'InstanceOfValidator' or 'InValidator'. attribute {0} does not.".format(
+                    attribute
+                )
             )
         return AttributeDetails(
             attribute.name, schema, is_required
         )
 
+    @classmethod
+    def _extract_attribute_schema_by_type(cls, extractor, attribute):
+        """
+        Extract the basic schema for the attribute based on its type.
+        The type can be supplied to two ways:
+        1. The type keyword of attr.ib
+        2. The _InstanceValidator of attr
+        """
+        schema = None
+        if attribute.type is not None:
+            schema = extractor.extract(attribute.type)
+        else:
+            for validator in _iterate_validator(attribute.validator):
+                if isinstance(validator, _InstanceOfValidator):
+                    schema = extractor.extract(validator.type)
+        return schema
+
+
 @attr.s
 class AttributeDetails(object):
     name = attr.ib()
     schema = attr.ib()
     is_required = attr.ib()
 
+
 def _iterate_validator(validator):
     if isinstance(validator, _AndValidator):
         for sub_validator in validator._validators:
             yield sub_validator
     else:
         yield validator
+
+
+def _enum_to_values(enum_type):
+    """
+    Convert an Enum into a list of values
+    """
+    return [enum_attribute.value for enum_attribute in list(enum_type)]
```

Comment:
nitpick: you don't need the list cast for enum_type. Lists can only convert iterables, which a list comprehension also supports.

Reply:
I guess technically this is beyond a nitpick, as the list cast can increase memory allocation for no benefit.